components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1267,29 +1268,34 @@ where
         );
 
         let msg_type = msg.get_message().get_msg_type();
-        let store_id = self.ctx.store_id();
 
-        if disk::disk_full_precheck(store_id) || self.ctx.is_disk_full {
-            let mut flag = false;
+        if disk::is_disk_almost_full(self.ctx.disk_status, self.ctx.store.get_id())
+            || disk::is_disk_already_full(self.ctx.disk_status, self.ctx.store.get_id())
+        {
+            // only leader transfer and winning log are allowed.
+            let mut allowed = true;
             if MessageType::MsgAppend == msg_type {
                 let entries = msg.get_message().get_entries();
                 for i in entries {
                     let entry_type = i.get_entry_type();
                     if EntryType::EntryNormal == entry_type && !i.get_data().is_empty() {
-                        flag = true;
+                        allowed = false;
                         break;
                     }
                 }
             } else if MessageType::MsgTimeoutNow == msg_type {
-                flag = true;
+                allowed = false;
             }
 
-            if flag {
+            if !allowed {
                 debug!(
                     "skip {:?} because of disk full", msg_type;
                     "region_id" => self.region_id(), "peer_id" => self.fsm.peer_id()
                 );
-                return Err(Error::Timeout("disk full".to_owned()));
+                return Err(Error::DiskFull(
```

Comment:
So the message is dropped. Please update the metrics `STORE_RAFT_DROPPED_MESSAGE_COUNTER`.